### ISSUE TITLE
Slow performance of AvroOutputStream compared to GenericRecord usage

### ISSUE THREAD

**Issue by RayRoestenburg** [sksamuel/avro4s#95]
Hi, 

I was surprised by how slow avro4s serializes compared to benchmarks shown in https://github.com/eishay/jvm-serializers/wiki and did a little searching. For my specific schema (which I cannot share), the below code would do about 7000 serializations / sec (repeatedly calling `serialize`).

```scala
def serialize[T: SchemaFor: ToRecord](data: T): Array[Byte] = {
  val baos = new ByteArrayOutputStream()
  val output = AvroOutputStream.binary[T](baos)
  try {
    output.write(data)
    output.flush()
    baos.toByteArray
  } finally output.close()
}
```

While dropping down to `GenericRecord` and reusing some of the avro objects was about 14 times faster, at max 100.000 serializations / sec. (I got about 120.000 serializations / sec using the avro SpecificRecord API directly with avrohugger generated code). 

```scala
class MyMessageSerializer {
  val encoderFactory = EncoderFactory.get()
  val schema = AvroSchema[MyMessage]
  val writer = new GenericDatumWriter[GenericRecord](schema)
  val format = RecordFormat[MyMessage]
  var encoder: BinaryEncoder = null

  def serialize(data: MyMessage): Array[Byte] = {
    val genericRecord = format.to(data)
    val out = new java.io.ByteArrayOutputStream()
    encoder = encoderFactory.binaryEncoder(out, encoder)
    writer.write(genericRecord, encoder)
    encoder.flush()
    out.toByteArray()
  }
}
```

Which seems like it's got to do with not reusing (one, some or all of) schema, writer, format and encoder across serializations. (It's similar for deserializing)

**Comment by sksamuel** [comment]
If you call serialize repeatedly, it will be re-calling the implicit method to generate the schema each time.
You could try bringing that in first,

```scala
implicit val schema = SchemaFor[X]
implicit val torecord = ToRecord[X]
records.foreach(out.serialize)
```

**Comment by RayRoestenburg** [comment]
Thanks, that fixes the issue, I get to around 100.000 ops /sec again with that. 

It's a bit obvious after the fact, but maybe useful to add to documentation that the implicit instances are not cached? (Coming from something like spray-json where you define implicits in an object you don't think about this)

**Comment by RayRoestenburg** [comment]
And thank you for the quick response!
